### PR TITLE
add `numel` method with default attribute

### LINKED ISSUE #240: `torch.Tensor.numel` method : Don't know how to interpret a callable with type <class 'int'>
It is because `TensorProxy` models `numel` as a property but it is a method in a PyTorch and numpy doesn't have `numel` but has `size`. I think this `numel` should be modelled based on current language context to be a method for PyTorch context and error in case of Numpy context.

```python
import thunder
import torch

import thunder.examine

def foo(x):
    return x.numel()

# thunder.examine.examine(foo, torch.rand(3, 3))
thunder.jit(foo)(torch.rand(3, 3))

```

Error:
```
 File "/home/kkalambarkar/lightning-thunder/thunder/core/interpreter.py", line 6664, in fn_
    raise InterpreterError(msg) from e
thunder.core.interpreter.InterpreterError: Encountered exception NotImplementedError: Don't know how to interpret a callable with type <class 'int'> without a __call__ method while tracing <function foo at 0x7f339910fd90>:
```

cc @apaz-cli

### PR DESCRIPTION
**Pull request by k223kim**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #240. 

I have modified the use of `numel` in a torch context to `numel()`. Also, I have added a test case to compare the default `numel` attribute and torch version of `numel`. There is a operator test cas in `opinfos.py`.
Also, the test that is failing for this PR seems irrelevant to the PR (it occured once I merged the `main` branch to this branch and the error is related to `test_phantom_grad_vs_torch_consistency_getitem_`).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Got to understand how language context is being interpreted and used! So much fun!

### CODE REVIEW

File: thunder/core/langctxs.py [hunk 1]
```diff
@@ -64,8 +64,17 @@ def reset_langctx(token: Any, /) -> None:
 
 # A helper for acquiring a method
 def resolve_method(id: Any, *args, **kwargs) -> Callable:
```
Comment (mruberry) [inline thunder/core/langctxs.py:66]: Update the return type to `None | Callable` since this may now return `None`

File: thunder/core/langctxs.py [hunk 2]
```diff
@@ -64,8 +64,17 @@ def reset_langctx(token: Any, /) -> None:
...
-    ctx: LanguageContext = get_langctx()
-    method: Callable = ctx.get_method(id, *args, **kwargs)
+    # I could combine the try/except statements but intentionally separated them for clarity
```
Comment (mruberry) [inline thunder/core/langctxs.py:67]: Let's remove this comment. It makes sense, but most readers won't be interested in this detail

File: thunder/core/langctxs.py [hunk 3]
```diff
@@ -64,8 +64,17 @@ def reset_langctx(token: Any, /) -> None:
...
+    try:
+        ctx: None | LanguageContext = get_langctx()
+    except LookupError:
+        return None
+    try:
+        # AttributeError occurs when there is language context is None or when that context does not have the attribute
```
Comment (mruberry) [inline thunder/core/langctxs.py:73]: This comment is a little confusing. If the language context is `None`, then won't that be caught by the previous case?

So should this now be:

```
# ctx.get_method throws an AttributeError when the context does not have the requested attribute, except
# for the prims language context, which always throws a ValueError
```

?

File: thunder/tests/opinfos.py [hunk 4]
```diff
@@ -4561,6 +4561,28 @@ def unsqueeze_sample_generator(op, device, dtype, requires_grad, **kwargs):
...
+        yield SampleInput(make(shape))
+
+
+numel_opinfo = OpInfo(
+    ltorch.numel,
+    dtypes=(datatypes.floating,),
+    sample_input_generator=numel_sample_generator,
+    torch_reference=torch.numel,
+)
+reduction_ops.append(numel_opinfo)
```
Comment (mruberry) [inline thunder/tests/opinfos.py:4583]: Let's put this OpInfo in `tensor_properties`:

https://github.com/k223kim/lightning-thunder/blob/020946dd99630e7b2c963087d5518844fd90643a/thunder/tests/opinfos.py#L548

File: thunder/tests/test_core.py [hunk 5]
```diff
@@ -2313,6 +2313,29 @@ def forward(self, x):
     assert "t_fc2_weight" in sig.parameters
 
 
+@instantiate(dtypes=(thunder.float32,))
+def test_default_method(executor, device: str, dtype: dtypes.dtype):
+    # This test ensures that when no language context is given, it will fallback to the default implementation.
```
Comment (mruberry) [inline thunder/tests/test_core.py:2318]: Great test description

File: thunder/tests/test_core.py [hunk 6]
```diff
@@ -2313,6 +2313,29 @@ def forward(self, x):
...
+    with detached_trace():
+        b = TensorProxy(
+            name="__b",
+            shape=(2, 2),
+            device=thunder.core.devices.cpu,
+            dtype=thunder.core.dtypes.float32,
+            requires_grad=False,
+        )
+
+    # torch.numel(a) and a.numel() will run on PyTorch contenxt
```
Comment (mruberry) [inline thunder/tests/test_core.py:2334]: Great comment; nice test

File: thunder/torch/__init__.py [hunk 7]
```diff
@@ -179,6 +179,18 @@ def fn_(idx: int | None = None):
     return fn_
 
 
+@torchsymbol(torch.numel, torch.Tensor.numel, is_method=True)
+def numel(a: TensorLike, /) -> Number:
```
Comment (mruberry) [inline thunder/torch/__init__.py:183]: Really nice signature

File: thunder/torch/__init__.py [hunk 8]
```diff
@@ -179,6 +179,18 @@ def fn_(idx: int | None = None):
...
+    size = a.size()
```
Comment (mruberry) [inline thunder/torch/__init__.py:184]: I think the implementation of this can just be:

```
return a._numel
```

**Review by mruberry — COMMENTED** [review]
Overall this looks great, @k223kim! I made some comments for your review.

fyi @IvanYashchuk and @nikitaved that this PR will make `.numel()` consistent with torch in the torch language context

File: thunder/core/langctxs.py (hunk 3 above)
Comment (k223kim) [reply thunder/core/langctxs.py:73]: Yes, you are right. Thanks for a clearer comment! Will update.

**Comment by k223kim** [comment]
Hi @mruberry! Thanks again for the suggestions and feedback :) Updated and resolved all comments just now (I've learned what kind of code comments are needed/useful) Let me know if I should do anything else regarding this PR!

**Review by mruberry — APPROVED** [review]
Thanks @k223kim! It's great to be consistent with PyTorch and extend our support for future changes like this, if needed

fyi @IvanYashchuk and @nikitaved for the updated `numel()` callsites.

**Comment by k223kim** [comment]
Hey @mruberry! Thanks for approving this PR. I can see that you have enabled auto merge, however, I think it has failed to merge as it is failing some CI testing which I think is irrelevant to this PR. Is there a way for me to bypass this? If I need to do some work on my end, I would be happy to do that :)

**Pull request merged**